tests/metrics/test_metric_lightning.py

```diff
@@ -192,3 +210,78 @@ def training_epoch_end(self, outputs):
     logged = trainer.logged_metrics
     assert torch.allclose(torch.tensor(logged["SumMetric_epoch"]), model.sum)
     assert torch.allclose(torch.tensor(logged["DiffMetric_epoch"]), model.diff)
+
+
+def test_log_metric_no_attributes_raises(tmpdir):
+
+    class TestModel(BoringModel):
+
+        def training_step(self, *args):
+            metric = SumMetric()
+            self.log("foo", metric)
+
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=1)
+    model = TestModel()
+    with pytest.raises(MisconfigurationException, match="Could not find the `LightningModule` attribute"):
+        trainer.fit(model)
+
+
+def test_log_metrics_wrong_attributes_raises(tmpdir):
+
+    class TestModel(BoringModel):
+
+        def __init__(self):
+            super().__init__()
+
+            self.a_metric = SumMetric()
+
+        def training_step(self, *args):
+            metric = SumMetric()
+            self.log("foo", metric)
+
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=1)
+    model = TestModel()
+    with pytest.raises(MisconfigurationException, match=escape("where `name` is one of ['a_metric']")):
+        trainer.fit(model)
```

Comment:
@carmocca why do we need this dependency on the attribute name in the module vs the key name used for publishing? if it's solely for restoration, could we think of other approaches? this style of metrics logging is a very common pattern for us

Reply:
Not @carmocca, but I'll take a shot at this - I don't think there is a dependency on key name for publishing. This test is failing because the metric being logged isn't the same instance as the `self.a_metric` metric that's an attribute in the module.

`test_log_metric_dict` below logs some metrics that have key name != attribute name and still passes

Reply:
@mooey5775 is correct.

This is to match the logged metric with the metric attribute in the `LightningModule` so the state can be restored.

> could we think of other approaches?

Open to ideas!